// @blocksuite/phasor/dist/elements/shape/shapes/triangle.js
import { StrokeStyle } from '../../../consts.js';
import { Utils } from '../../../utils/tl-utils.js';
export const TriangleMethods = {
    render(ctx, rc, element) {
        const { w, h, seed, strokeWidth, filled, realFillColor, realStrokeColor, strokeStyle, roughness, } = element;
        const renderOffset = Math.max(strokeWidth, 0) / 2;
        const renderWidth = w - renderOffset * 2;
        const renderHeight = h - renderOffset * 2;
        ctx.translate(renderOffset, renderOffset);
        rc.polygon([
            [renderWidth / 2, 0],
            [renderWidth, renderHeight],
            [0, renderHeight],
        ], {
            seed,
            roughness,
            strokeLineDash: strokeStyle === StrokeStyle.Dashed ? [12, 12] : undefined,
            stroke: realStrokeColor,
            strokeWidth,
            fill: filled ? realFillColor : undefined,
        });
    },
    hitTest(x, y, bound, options) {
        const points = [
            [bound.x + bound.w / 2, bound.y + 0],
            [bound.x + bound.w, bound.y + bound.h],
            [bound.x + 0, bound.y + bound.h],
        ];
        return Utils.pointInPolygon([x, y], points);
    },
};
