// @blocksuite/blocks/dist/page-block/edgeless/toolbar/connector-tool/connector-tool-button.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import '../../components/tool-icon-button.js';
import './connector-menu.js';
import { ConnectorIcon } from '@blocksuite/global/config';
import { ConnectorMode } from '@blocksuite/phasor';
import { assertExists } from '@blocksuite/store';
import { createPopper } from '@popperjs/core';
import { css, html, LitElement } from 'lit';
import { customElement, property } from 'lit/decorators.js';
import { DEFAULT_SELECTED_COLOR } from '../../components/color-panel.js';
function createConnectorMenuPopper(reference) {
    const menu = document.createElement('edgeless-connector-menu');
    assertExists(reference.shadowRoot);
    reference.shadowRoot.appendChild(menu);
    const popper = createPopper(reference, menu, {
        placement: 'top',
        modifiers: [
            {
                name: 'offset',
                options: {
                    offset: [0, 12],
                },
            },
        ],
    });
    return {
        element: menu,
        dispose: () => {
            menu.remove();
            popper.destroy();
        },
    };
}
let EdgelessConnectorToolButton = class EdgelessConnectorToolButton extends LitElement {
    constructor() {
        super(...arguments);
        this._menu = null;
    }
    _toggleMenu() {
        if (this._menu) {
            this._menu.dispose();
            this._menu = null;
        }
        else {
            this._menu = createConnectorMenuPopper(this);
            this._menu.element.mouseMode = this.mouseMode;
            this._menu.element.edgeless = this.edgeless;
        }
    }
    updated(changedProperties) {
        if (changedProperties.has('mouseMode')) {
            if (this.mouseMode.type !== 'connector') {
                this._menu?.dispose();
                this._menu = null;
            }
            if (this._menu) {
                this._menu.element.mouseMode = this.mouseMode;
                this._menu.element.edgeless = this.edgeless;
            }
        }
    }
    disconnectedCallback() {
        this._menu?.dispose();
        this._menu = null;
        super.disconnectedCallback();
    }
    render() {
        const type = this.mouseMode?.type;
        return html `
      <edgeless-tool-icon-button
        .tooltip=${'Connector'}
        .active=${type === 'connector'}
        @click=${() => {
            this.setMouseMode({
                type: 'connector',
                mode: ConnectorMode.Orthogonal,
                color: DEFAULT_SELECTED_COLOR,
            });
            this._toggleMenu();
        }}
      >
        ${ConnectorIcon}
      </edgeless-tool-icon-button>
    `;
    }
};
EdgelessConnectorToolButton.styles = css `
    :host {
      display: flex;
    }
  `;
__decorate([
    property()
], EdgelessConnectorToolButton.prototype, "mouseMode", void 0);
__decorate([
    property()
], EdgelessConnectorToolButton.prototype, "edgeless", void 0);
__decorate([
    property()
], EdgelessConnectorToolButton.prototype, "setMouseMode", void 0);
EdgelessConnectorToolButton = __decorate([
    customElement('edgeless-connector-tool-button')
], EdgelessConnectorToolButton);
export { EdgelessConnectorToolButton };
