// @blocksuite/global/dist/utils/slot.js
import { flattenDisposables } from './disposable.js';
// Credits to blocky-editor
// https://github.com/vincentdchan/blocky-editor
export class Slot {
    constructor() {
        this._emitting = false;
        this._callbacks = [];
        this._disposables = [];
        this.subscribe = (selector, callback, config) => {
            let prevState;
            const { filter, equalityFn = Object.is } = config ?? {};
            return this.on(state => {
                if (filter && !filter(state)) {
                    return;
                }
                const nextState = selector(state);
                if (prevState === undefined || !equalityFn(prevState, nextState)) {
                    callback(nextState);
                    prevState = nextState;
                }
            });
        };
    }
    filter(testFun) {
        const result = new Slot();
        // if result is disposed, dispose this too
        result._disposables.push({ dispose: () => this.dispose() });
        this.on((v) => {
            if (testFun(v)) {
                result.emit(v);
            }
        });
        return result;
    }
    on(callback) {
        if (this._emitting) {
            const newCallback = [...this._callbacks, callback];
            this._callbacks = newCallback;
        }
        else {
            this._callbacks.push(callback);
        }
        return {
            dispose: () => {
                if (this._emitting) {
                    this._callbacks = this._callbacks.filter(v => v !== callback);
                }
                else {
                    const index = this._callbacks.indexOf(callback);
                    if (index > -1) {
                        this._callbacks.splice(index, 1); // remove one item only
                    }
                }
            },
        };
    }
    once(callback) {
        let dispose = undefined;
        const handler = (v) => {
            callback(v);
            if (dispose) {
                dispose();
            }
        };
        const disposable = this.on(handler);
        dispose = disposable.dispose;
        return disposable;
    }
    unshift(callback) {
        if (this._emitting) {
            const newCallback = [callback, ...this._callbacks];
            this._callbacks = newCallback;
        }
        else {
            this._callbacks.unshift(callback);
        }
        return {
            dispose: () => {
                if (this._emitting) {
                    this._callbacks = this._callbacks.filter(v => v !== callback);
                }
                else {
                    const index = this._callbacks.indexOf(callback);
                    if (index > -1) {
                        this._callbacks.splice(index, 1); // remove one item only
                    }
                }
            },
        };
    }
    emit(v) {
        const prevEmitting = this._emitting;
        this._emitting = true;
        this._callbacks.forEach(f => {
            try {
                f(v);
            }
            catch (err) {
                console.error(err);
            }
        });
        this._emitting = prevEmitting;
    }
    pipe(that) {
        this._callbacks.push(v => that.emit(v));
        return this;
    }
    dispose() {
        flattenDisposables(this._disposables).dispose();
        this._callbacks = [];
        this._disposables = [];
    }
    toDispose(disposables) {
        disposables.push(this);
        return this;
    }
}
