// @blocksuite/blocks/dist/page-block/edgeless/components/line-styles-panel.js
import { html } from 'lit';
import { repeat } from 'lit/directives/repeat.js';
import { LineSizeButton, lineSizeButtonStyles, } from './line-size-button.js';
import { LineStyleButton, lineStyleButtonStyles, } from './line-style-button.js';
import { PanelWrapper, panelWrapperStyle } from './panel-wrapper.js';
export const lineStylesPanelStyles = [
    panelWrapperStyle,
    lineSizeButtonStyles,
    lineStyleButtonStyles,
];
export function LineStylesPanel({ onClick, selectedLineSize, selectedLineStyle, lineStyle = ['solid', 'dash', 'none'], } = {}) {
    const lineSizeButtons = repeat(['s', 'l'], size => size, size => {
        return LineSizeButton({
            size,
            active: size === selectedLineSize,
            onClick: () => {
                onClick?.({
                    type: 'size',
                    value: size,
                });
            },
        });
    });
    const lineStyleButtons = repeat(lineStyle, mode => mode, mode => {
        return LineStyleButton({
            mode,
            active: mode === selectedLineStyle,
            onClick: () => {
                onClick?.({
                    type: 'lineStyle',
                    value: mode,
                });
            },
        });
    });
    return PanelWrapper({
        className: 'line-style-panel',
        children: html `
      ${lineSizeButtons}
      <menu-divider .vertical=${true}></menu-divider>
      ${lineStyleButtons}
    `,
    });
}
