// @blocksuite/blocks/dist/page-block/edgeless/components/line-style-button.js
import './tool-icon-button.js';
import { BanIcon, ConnectorLIcon, DashLineIcon, } from '@blocksuite/global/config';
import { css, html } from 'lit';
export const lineStyleButtonStyles = [
    css `
    .edgeless-component-line-style-button {
      display: flex;
      justify-content: center;
      align-items: center;
      width: 16px;
      height: 16px;
      color: var(--affine-icon-color);
    }
  `,
];
function getModeIcon(mode) {
    switch (mode) {
        case 'solid': {
            return ConnectorLIcon;
        }
        case 'dash': {
            return DashLineIcon;
        }
        case 'none': {
            return BanIcon;
        }
    }
}
function getModeTooltip(mode) {
    return {
        solid: 'Solid',
        dash: 'Dash',
        none: 'None',
    }[mode];
}
export function LineStyleButton({ className, mode, active, tooltip: tooltipFromProps, onClick, }) {
    const classnames = `edgeless-component-line-style-button mode-${mode} ${active ? 'active' : ''}`;
    const icon = getModeIcon(mode);
    const tooltip = tooltipFromProps !== undefined ? tooltipFromProps : getModeTooltip(mode);
    return html `<edgeless-tool-icon-button
    class=${className}
    .active=${active}
    .activeMode=${'background'}
    .tooltip=${tooltip}
    @click=${onClick}
  >
    <div class=${classnames}>${icon}</div>
  </edgeless-tool-icon-button>`;
}
