// @blocksuite/phasor/dist/elements/shape/shape-element.js
import { SurfaceElement } from '../surface-element.js';
import { ShapeMethodsMap } from './shapes/index.js';
export class ShapeElement extends SurfaceElement {
    get shapeType() {
        const shapeType = this.yMap.get('shapeType');
        return shapeType;
    }
    get radius() {
        const radius = this.yMap.get('radius');
        return radius;
    }
    get filled() {
        const filled = this.yMap.get('filled');
        return filled;
    }
    get fillColor() {
        const fillColor = this.yMap.get('fillColor');
        return fillColor;
    }
    get strokeWidth() {
        const strokeWidth = this.yMap.get('strokeWidth');
        return strokeWidth;
    }
    get strokeColor() {
        const strokeColor = this.yMap.get('strokeColor');
        return strokeColor;
    }
    get strokeStyle() {
        const strokeStyle = this.yMap.get('strokeStyle');
        return strokeStyle;
    }
    get roughness() {
        const roughness = this.yMap.get('roughness') ?? 2;
        return roughness;
    }
    get realStrokeColor() {
        return this.computedValue(this.strokeColor);
    }
    get realFillColor() {
        return this.computedValue(this.fillColor);
    }
    hitTest(x, y, options) {
        const { hitTest } = ShapeMethodsMap[this.shapeType];
        return hitTest(x, y, this, options);
    }
    render(ctx, rc) {
        const { render } = ShapeMethodsMap[this.shapeType];
        render(ctx, rc, this);
    }
}
