// @blocksuite/blocks/dist/embed-block/image/placeholder/circle-loading.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { ShadowlessElement } from '@blocksuite/lit';
import { css, html } from 'lit';
import { customElement } from 'lit/decorators.js';
const ELEMENT_TAG = 'affine-image-block-circle-loading';
let AffineImageBlockCircleLoading = class AffineImageBlockCircleLoading extends ShadowlessElement {
    render() {
        return html `<div class="affine-image-block-container">
      <div class="affine-image-block-loading"></div>
    </div>`;
    }
};
AffineImageBlockCircleLoading.styles = css `
    @keyframes affine-image-block-rotate {
      from {
        rotate: 0deg;
      }
      to {
        rotate: 360deg;
      }
    }

    .affine-image-block-container {
      width: 24px;
      height: 24px;
      overflow: hidden;
    }

    .affine-image-block-loading {
      display: inline-block;
      width: 24px;
      height: 24px;
      position: relative;
      background: conic-gradient(rgba(255, 255, 255, 0.31), #6880ff);
      border-radius: 50%;
      animation: affine-image-block-rotate 1s infinite ease-in;
    }

    .affine-image-block-loading::before {
      content: '';
      width: 20px;
      height: 20px;
      border-radius: 50%;
      background-color: white;
      position: absolute;
      top: 2px;
      left: 2px;
    }
  `;
AffineImageBlockCircleLoading = __decorate([
    customElement(ELEMENT_TAG)
], AffineImageBlockCircleLoading);
export { AffineImageBlockCircleLoading };
