// @blocksuite/blocks/dist/embed-block/image/placeholder/loading-card.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import './circle-loading.js';
import { ShadowlessElement } from '@blocksuite/lit';
import { css, html } from 'lit';
import { customElement, property } from 'lit/decorators.js';
const ELEMENT_TAG = 'affine-image-block-loading-card';
let AffineImageBlockLoadingCard = class AffineImageBlockLoadingCard extends ShadowlessElement {
    constructor() {
        super(...arguments);
        this.content = 'Loading content...';
    }
    render() {
        return html `
      <div class="affine-image-block-loading-card">
        <affine-image-block-circle-loading></affine-image-block-circle-loading>
        <div class="affine-image-block-content">${this.content}</div>
      </div>
    `;
    }
};
AffineImageBlockLoadingCard.styles = css `
    .affine-image-block-loading-card {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      width: 100%;
      height: 100%;
      margin: 0 auto;
      border: 1px solid #ebeeff;
      border-radius: 10px;
      background: #fbfbff;
    }

    .affine-image-block-content {
      height: 30px;
      line-height: 22px;
      padding-top: 8px;
      color: var(--affine-primary-color);
      font-size: 16px;
      font-weight: 400;
    }
  `;
__decorate([
    property()
], AffineImageBlockLoadingCard.prototype, "content", void 0);
AffineImageBlockLoadingCard = __decorate([
    customElement(ELEMENT_TAG)
], AffineImageBlockLoadingCard);
export { AffineImageBlockLoadingCard };
