// @blocksuite/phasor/dist/elements/index.js
import { BrushElement } from './brush/brush-element.js';
import { BrushElementDefaultProps } from './brush/constants.js';
import { ConnectorElement } from './connector/connector-element.js';
import { ConnectorElementDefaultProps } from './connector/constants.js';
import { DebugElement, DebugElementDefaultProps, } from './debug/debug-element.js';
import { ShapeElementDefaultProps } from './shape/constants.js';
import { ShapeElement } from './shape/shape-element.js';
import { TextElementDefaultProps } from './text/constants.js';
import { TextElement } from './text/text-element.js';
export { BrushElement } from './brush/brush-element.js';
export { ConnectorElement } from './connector/connector-element.js';
export { DebugElement } from './debug/debug-element.js';
export { ShapeElement } from './shape/shape-element.js';
export { TextElement } from './text/text-element.js';
export const ElementCtors = {
    debug: DebugElement,
    brush: BrushElement,
    shape: ShapeElement,
    connector: ConnectorElement,
    text: TextElement,
};
export const ElementDefaultProps = {
    debug: DebugElementDefaultProps,
    brush: BrushElementDefaultProps,
    shape: ShapeElementDefaultProps,
    connector: ConnectorElementDefaultProps,
    text: TextElementDefaultProps,
};
