// @blocksuite/blocks/dist/frame-block/frame-service.js
import { getService } from '../__internal__/service.js';
import { BaseService } from '../__internal__/service/index.js';
import { addSerializedBlocks } from '../__internal__/service/json2block.js';
export class FrameBlockService extends BaseService {
    async json2Block(focusedBlockModel, pastedBlocks) {
        await addSerializedBlocks(focusedBlockModel.page, pastedBlocks, focusedBlockModel, 0);
    }
    block2Json(block, begin, end) {
        const delta = block.text?.sliceToDelta(begin || 0, end) || [];
        return {
            flavour: block.flavour,
            type: block.type,
            text: delta,
            xywh: block.xywh,
            background: block.background,
            children: block.children?.map((child, index) => {
                if (index === block.children.length - 1) {
                    return getService(child.flavour).block2Json(child, 0, end);
                }
                return getService(child.flavour).block2Json(child);
            }),
        };
    }
}
