// @blocksuite/blocks/dist/database-block/table/components/column-header/column-move/column-drag-preview.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { ShadowlessElement } from '@blocksuite/lit';
import { html } from 'lit';
import { customElement } from 'lit/decorators.js';
let ColumnDragPreview = class ColumnDragPreview extends ShadowlessElement {
    render() {
        return html `
      <style>
        affine-database-column-drag-preview {
          display: flex;
          flex-direction: column;
          position: fixed;
          top: 0;
          left: 0;
          height: 220px;
          width: fit-content;
          border: 1px solid var(--affine-border-color);
          border-radius: 4px;
          overflow: hidden;
          cursor: none;
          user-select: none;
          pointer-events: none;
          caret-color: transparent;
          z-index: 100;
        }

        .preview-column-header {
          opacity: 0.8;
          border-bottom: 1px solid var(--affine-border-color);
          background: var(--affine-tertiary-color);
        }
        .preview-column-header .affine-database-column-move svg {
          cursor: grabbing;
          opacity: 1;
        }
        .preview-column-header .affine-database-column-move circle {
          fill: var(--affine-text-emphasis-color);
        }

        .preview-column-content {
          flex: 1;
          opacity: 0.8;
          background: var(--affine-white);
        }
      </style>
    `;
    }
};
ColumnDragPreview = __decorate([
    customElement('affine-database-column-drag-preview')
], ColumnDragPreview);
export { ColumnDragPreview };
