// @blocksuite/blocks/dist/page-block/edgeless/toolbar/brush-tool/brush-menu.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import '../../components/color-panel.js';
import '../../components/tool-icon-button.js';
import { css, html, LitElement, nothing } from 'lit';
import { customElement, property } from 'lit/decorators.js';
import { BrushSize } from '../../../../__internal__/index.js';
import { tooltipStyle } from '../../../../components/tooltip/tooltip.js';
function BrushSizeButtonGroup(mouseMode, setBrushWidth) {
    if (mouseMode.type !== 'brush')
        return nothing;
    const { lineWidth } = mouseMode;
    /**
     * There is little hacky on rendering tooltip.
     * We don't want either tooltip overlap the top button or tooltip on left.
     * So we put the lower button's tooltip as the first element of the button group container
     */
    return html `
    <div class="brush-size-button-group has-tool-tip">
      <!-- This tooltip is for the last button(Thick) -->
      <tool-tip inert role="tooltip" tip-position="top" arrow>Thick</tool-tip>

      <div
        class="brush-size-button has-tool-tip"
        ?active=${lineWidth === BrushSize.Thin}
        @click=${() => setBrushWidth(BrushSize.Thin)}
      >
        <div class="thin"></div>
        <tool-tip inert role="tooltip" tip-position="top" arrow>
          Thin
        </tool-tip>
      </div>

      <div
        class="brush-size-button"
        ?active=${lineWidth === BrushSize.Thick}
        @click=${() => setBrushWidth(BrushSize.Thick)}
      >
        <div class="thick"></div>
      </div>
    </div>
  `;
}
let EdgelessBrushMenu = class EdgelessBrushMenu extends LitElement {
    constructor() {
        super(...arguments);
        this._setBrushColor = (color) => {
            if (this.mouseMode.type !== 'brush')
                return;
            const { lineWidth } = this.mouseMode;
            this.edgeless.slots.mouseModeUpdated.emit({
                type: 'brush',
                color,
                lineWidth,
            });
        };
        this._setBrushWidth = (lineWidth) => {
            if (this.mouseMode.type !== 'brush')
                return;
            const { color } = this.mouseMode;
            this.edgeless.slots.mouseModeUpdated.emit({
                type: 'brush',
                color,
                lineWidth,
            });
        };
    }
    render() {
        if (this.mouseMode.type !== 'brush')
            return nothing;
        const { color } = this.mouseMode;
        const brushSizeButtonGroup = BrushSizeButtonGroup(this.mouseMode, this._setBrushWidth);
        return html `
      <div class="container">
        ${brushSizeButtonGroup}
        <menu-divider .vertical=${true}></menu-divider>
        <edgeless-color-panel
          .value=${color}
          @select=${(e) => this._setBrushColor(e.detail)}
        ></edgeless-color-panel>
      </div>
    `;
    }
};
EdgelessBrushMenu.styles = css `
    :host {
      width: 260px;
      z-index: 1;
    }
    .container {
      display: flex;
      padding: 4px;
      justify-content: center;
      align-items: center;
      background: var(--affine-background-overlay-panel-color);
      box-shadow: var(--affine-shadow-2);
      border-radius: 8px;
    }

    .brush-size-button-group {
      display: flex;
      flex-direction: column;
    }

    .brush-size-button {
      display: flex;
      justify-content: center;
      align-items: center;
      width: 32px;
      height: 32px;
      box-sizing: border-box;
      border-radius: 4px;
      cursor: pointer;
    }

    .brush-size-button[active],
    .brush-size-button:hover {
      background-color: var(--affine-hover-color);
    }

    .brush-size-button div {
      border-radius: 50%;
      background-color: var(--affine-icon-color);
    }

    .brush-size-button .thin {
      width: 4px;
      height: 4px;
    }

    .brush-size-button .thick {
      width: 10px;
      height: 10px;
    }

    menu-divider {
      height: 62px;
    }

    ${tooltipStyle}
  `;
__decorate([
    property()
], EdgelessBrushMenu.prototype, "mouseMode", void 0);
__decorate([
    property()
], EdgelessBrushMenu.prototype, "edgeless", void 0);
EdgelessBrushMenu = __decorate([
    customElement('edgeless-brush-menu')
], EdgelessBrushMenu);
export { EdgelessBrushMenu };
